### PR TITLE
Sort students by last, then first name

### PR DESCRIPTION
**Pull request by LMS007**
The client attempts to discover a last name based off the displayName, but if it does not exist, then it sorts by first name.

Additionally if there are more separators, then every word past the first is assumed to be part of the last name. This may be a false assumption but its a best effort with only having a displayName.

Finally, the displayName is changed from the service layer to show the last then first. e.g. “{last}, {first}”

---------
I thought this was going to be a bit more straight forward, nothing with LMS ever is! :)
So we have to make some assumptions here and handle all the edge cases. We really don't know what the first or last names are so I went ahead with some general assumptions that hopefully are mostly right or at least make some sense when ordering.

To test, try this assignment.
https://blackboard.hypothes.is/webapps/blackboard/execute/blti/launchLink?course_id=_16_1&content_id=_214_1

It only has 2 students because we only have 2 student accounts, but the unit test should have all the edge cases

students

fixes https://github.com/hypothesis/lms/issues/1629

### CODE REVIEW

File: lms/static/scripts/frontend_apps/components/LMSGrader.js [hunk 1]
```diff
@@ -17,11 +17,97 @@ export default function LMSGrader({
...
+    // so we can compare on those values.
+    const students_ = unorderedStudents.map(s => {
+      return {
+        ...s,
+        ...makeNames(s.displayName),
+      };
+    });
+    students_.sort((student1, student2) => {
+      function compareNames(name1, name2) {
+        if (name1 < name2) {
```
Comment (LMS007) [inline lms/static/scripts/frontend_apps/components/LMSGrader.js:83]: Case will also matter since its not converting, but I figured that was okay?

**Comment by lyzadanger** [comment]
I've been burned in the past and am thus gun shy about anything to do with parsing people's names and trying to be smart. It's a non-trivial problem. Casing is just the tip of the iceberg.

My gut feeling is that we should try to avoid doing anything clever at all, even if that means literally just alphabetizing by full `displayName` or fallback field. That would mean no parsing or splitting.

Consider these names (all real humans except for one that I disguised slightly for privacy) and what would happen to them with these changes (in parentheses):

* E Ben Sanderson (Ben Sanderson, E) - This person's surname is "Sanderson", not "Ben Sanderson"
* Aaron ben Ze-ev (ben Ze-ev, Aaron) - This is "correct", but note that it is the same pattern as the above.
* Lyza Danger Gardner (Danger Gardner, Lyza) - My last name is not Danger
* E. B. White - (B. White, E.)
* Lev Nikolayevich Tolstoy - (Nikolayevich Tolstoy, Lev) Nikolayevich is a patronymic, not a surname
* Tunjung - (Tunjung) This person uses one name. These changes would get it "correct", but I wanted point it out anyway.
* Bong Joon-ho - (Joon-ho, Bong) Bong is this person's family name/surname (Korean).

Cultures worldwide have different naming patterns and order given names/surnames differently, and some cultures don't use surnames at all. People have 0 - n middle names.

> So we have to make some assumptions here and handle all the edge cases

I'd gently argue against this; it'll be an arms war and we're not going to catch everything.

File: lms/static/scripts/frontend_apps/components/LMSGrader.js [hunk 2]
```diff
@@ -17,11 +17,31 @@ export default function LMSGrader({
...
   courseName,
-  students,
+  students: unorderedStudents,
 }) {
   // No initial current student selected
   const [currentStudentIndex, setCurrentStudentIndex] = useState(-1);
 
+  // A sorted list of students. Students are sorted by
+  // displayName.
+  const [students] = useState(() => {
```
Comment (lyzadanger) [inline lms/static/scripts/frontend_apps/components/LMSGrader.js:27]: I think `useMemo` may work here, but `useState`, as we discussed in Slack, is probably not whatcha want.

File: lms/static/scripts/frontend_apps/components/LMSGrader.js [hunk 3]
```diff
@@ -17,11 +17,31 @@ export default function LMSGrader({
...
+    // Make a copy
+    const students_ = [...unorderedStudents];
+    students_.sort((student1, student2) => {
+      function compareNames(name1, name2) {
```
Comment (lyzadanger) [inline lms/static/scripts/frontend_apps/components/LMSGrader.js:31]: I think it wise to define this function outside of the `Array.prototype.sort` call.

File: lms/static/scripts/frontend_apps/components/LMSGrader.js [hunk 4]
```diff
@@ -17,11 +17,31 @@ export default function LMSGrader({
...
+        if (name1.toLowerCase() < name2.toLowerCase()) {
+          return -1;
+        } else if (name1.toLowerCase() > name2.toLowerCase()) {
+          return 1;
+        } else {
+          return 0;
+        }
+      }
+      return compareNames(student1.displayName, student2.displayName);
```
Comment (lyzadanger) [inline lms/static/scripts/frontend_apps/components/LMSGrader.js:40]: Are we 100% sure we always have a `displayName`? Maybe we are!

File: lms/static/scripts/frontend_apps/components/test/LMSGrader-test.js [hunk 5]
```diff
@@ -81,6 +82,52 @@ describe('LMSGrader', () => {
...
+      .map(s => {
+        return s.displayName;
+      });
+    assert.match(
+      [
+        'Alpha',
+        'Beta',
+        'Student Alpha',
+        'Student Beta',
+        'student Beta',
```
Comment (lyzadanger) [inline lms/static/scripts/frontend_apps/components/test/LMSGrader-test.js:123]: Is it simply because it comes later in the unsorted `Array` that 'student Beta' ends up after 'Student Beta'?

File: lms/static/scripts/frontend_apps/components/LMSGrader.js (hunk 4 above)
Comment (LMS007) [reply lms/static/scripts/frontend_apps/components/LMSGrader.js:40]: Not likley, but its easy to make at least this code 
1. Not break
2. Sort it first

It would likely be a bad state for the app, but we can also attempt to make it less bad by not throwing confusing JS errors.

File: lms/static/scripts/frontend_apps/components/test/LMSGrader-test.js (hunk 5 above)
Comment (LMS007) [reply lms/static/scripts/frontend_apps/components/test/LMSGrader-test.js:123]: correct, these are a tie, or its initial order actually matters. I'll add a comment

File: lms/static/scripts/frontend_apps/components/test/LMSGrader-test.js [hunk 6]
```diff
@@ -81,6 +97,97 @@ describe('LMSGrader', () => {
     );
   });
 
+  it('orders the students by last name (if it exists), then first name', () => {
```
Comment (lyzadanger) [inline lms/static/scripts/frontend_apps/components/test/LMSGrader-test.js:100]: Is this test description out of date?

File: lms/static/scripts/frontend_apps/components/test/LMSGrader-test.js [hunk 7]
```diff
@@ -81,6 +97,97 @@ describe('LMSGrader', () => {
...
+        {
+          displayName: 'Gamma',
+        },
+      ],
+    });
+    orderedStudentNames = getDisplayNames(wrapper);
+    assert.match(['Beta', 'Gamma'], orderedStudentNames);
+  });
+
+  it('sorts an undefined displayName first', () => {
```
Comment (lyzadanger) [inline lms/static/scripts/frontend_apps/components/test/LMSGrader-test.js:171]: Perhaps this could be worded as "puts students with empty `displayName`s at the beginning of sorted students"

File: lms/static/scripts/frontend_apps/components/LMSGrader.js [hunk 8]
```diff
@@ -17,11 +17,32 @@ export default function LMSGrader({
   children,
   assignmentName,
   courseName,
-  students,
+  students: unorderedStudents,
 }) {
   // No initial current student selected
   const [currentStudentIndex, setCurrentStudentIndex] = useState(-1);
 
+  // A sorted list of students. Students are sorted by
```
Comment (lyzadanger) [inline lms/static/scripts/frontend_apps/components/LMSGrader.js:25]: This comment could probably be made more concise: "Students sorted by `displayName`" or similar.

File: lms/static/scripts/frontend_apps/components/test/LMSGrader-test.js (hunk 6 above)
Comment (LMS007) [reply lms/static/scripts/frontend_apps/components/test/LMSGrader-test.js:100]: indeed

File: lms/static/scripts/frontend_apps/components/LMSGrader.js (hunk 8 above)
Comment (LMS007) [reply lms/static/scripts/frontend_apps/components/LMSGrader.js:25]: agreed
